posthog/queries/paths/paths.py

```diff
@@ -0,0 +1,397 @@
+import dataclasses
+from collections import defaultdict
+from re import escape
+from typing import Dict, List, Literal, Optional, Tuple, Union, cast
+
+from rest_framework.exceptions import ValidationError
+
+from ee.clickhouse.materialized_columns.columns import ColumnName
+from ee.clickhouse.queries.funnels.funnel_persons import ClickhouseFunnelActors
+from posthog.client import sync_execute
+from posthog.constants import FUNNEL_PATH_BETWEEN_STEPS, LIMIT, PATH_EDGE_LIMIT
+from posthog.models import Filter, Team
+from posthog.models.filters.path_filter import PathFilter
+from posthog.models.property import PropertyName
+from posthog.queries.paths.paths_event_query import PathEventQuery
+from posthog.queries.paths.sql import PATH_ARRAY_QUERY
+
+EVENT_IN_SESSION_LIMIT_DEFAULT = 5
+SESSION_TIME_THRESHOLD_DEFAULT = 1800000  # milliseconds to 30 minutes
+EDGE_LIMIT_DEFAULT = 50
+
+
+@dataclasses.dataclass
+class ExtraEventClauses:
+    final_select_statements: str
+    joined_path_tuple_select_statements: str
+    array_filter_select_statements: str
+    limited_path_tuple_elements: str
+    path_time_tuple_select_statements: str
+    paths_tuple_elements: str
+    group_array_select_statements: str
+
+
+class Paths:
+    event_query = PathEventQuery
+    _filter: PathFilter
+    _funnel_filter: Optional[Filter]
+    _team: Team
+    _extra_event_fields: List[ColumnName]
+    _extra_event_properties: List[PropertyName]
+
+    def __init__(self, filter: PathFilter, team: Team, funnel_filter: Optional[Filter] = None,) -> None:
+        self._filter = filter
+        self._team = team
+        self.params = {
+            "team_id": self._team.pk,
+            "event_in_session_limit": self._filter.step_limit or EVENT_IN_SESSION_LIMIT_DEFAULT,
+            "session_time_threshold": SESSION_TIME_THRESHOLD_DEFAULT,
+            "groupings": self._filter.path_groupings or None,
+            "regex_groupings": None,
+        }
+        self._funnel_filter = funnel_filter
+
+        self._extra_event_fields: List[ColumnName] = []
+        self._extra_event_properties: List[PropertyName] = []
+        if self._filter.include_recordings:
+            self._extra_event_fields = ["uuid", "timestamp"]
+            self._extra_event_properties = ["$session_id", "$window_id"]
+
+        if self._filter.include_all_custom_events and self._filter.custom_events:
+            raise ValidationError("Cannot include all custom events and specific custom events in the same query")
+
+        if not self._filter.limit:
+            self._filter = self._filter.with_data({LIMIT: 100})
+
+        if self._filter.path_groupings:
```

Comment:
Are we going a different route with paths, and not separating the ee functionality here?

Reply:
Good point. Thoughts on merging initial code movement and then followup with more FOSS refactoring? This should prevent unexpected regressions as these PRs will be mainly responsible for just moving code from one folder to another